lib/notifications/db/detour.ex

```diff
@@ -33,4 +43,105 @@ defmodule Notifications.Db.Detour do
       :status
     ])
   end
+
+  defmodule Queries do
+    @moduledoc """
+    Defines composable queries for retrieving `Notifications.Db.Detour` info.
+    """
+
+    import Ecto.Query
+
+    @doc """
+    The "base" query that queries `Notifications.Db.Detour`'s without restriction
+
+
+    ## Examples
+
+    The `base` query returns all Detour Notifications
+
+        iex> :detour
+        ...> |> insert()
+        ...> |> Notifications.Notification.create_activated_detour_notification_from_detour()
+        ...>
+        iex> all_detour_notifications =
+        ...>   Notifications.Db.Detour.Queries.base()
+        ...>   |> Skate.Repo.all()
+        ...>
+        iex> match?(
+        ...>   [
+        ...>     %Notifications.Db.Detour{
+        ...>     }
+        ...>   ],
+        ...>   all_detour_notifications
+        ...> )
+        true
+
+    """
+    def base() do
+      from(d in Notifications.Db.Detour, as: :detour_notification, select_merge: d)
+    end
+
+    @doc """
+    Retrieves detour information for notifications from the `Notifications.Db.Detour` table
+
+    ## Examples
+
+        iex> :detour
+        ...> |> insert()
+        ...> |> Notifications.Notification.create_activated_detour_notification_from_detour()
+        ...>
+        iex> all_detour_notifications =
+        ...>   Notifications.Db.Detour.Queries.get_derived_info()
+        ...>   |> Skate.Repo.all()
+        ...>
+        iex> [
+        ...>   %Notifications.Db.Detour{
+        ...>    route: route,
+        ...>    origin: origin,
+        ...>    headsign: headsign,
+        ...>    direction: direction
+        ...>   }
+        ...> ] = all_detour_notifications
+        iex> Enum.any?([route, origin, headsign, direction], &is_nil/1)
+        false
+
+    """
+    def get_derived_info(query \\ base()) do
+      from(
+        [detour_notification: dn] in query,
+        left_join: ad in assoc(dn, :detour),
+        as: :associated_detour,
+        select_merge: %{
+          route: ad.state["context"]["route"]["name"],
+          origin: ad.state["context"]["routePattern"]["name"],
+          headsign: ad.state["context"]["routePattern"]["headsign"],
+
+          # Ecto can't figure out how to index a JSON map via another JSON value
```

Comment:
**Praise** This explanation is very helpful for this query, both for what that `CAST` is doing and for why it's there in the first place.